File: dentsys-api/src/models/patient-model.js
import pool from '../../config/db.js';

export default class Patient {
    constructor(id, firstName, lastName, middleName, birthDate, age, sex, nickname, nationality, religion, occupation, referrer, reason, totalBalance) {
        this.id = id;
        this.firstName = firstName;
        this.lastName = lastName;
        this.middleName = middleName;
        this.birthDate = birthDate;
        this.age = age;
        this.sex = sex;
        this.nickname = nickname;
        this.nationality = nationality;
        this.religion = religion;
        this.occupation = occupation;
        this.referrer = referrer;
        this.reason = reason;
        this.totalBalance = totalBalance;
    }

    static async createPatient(data) {
        const { firstName, lastName, middleName, birthDate, age, sex, nickname, nationality, religion, occupation, referrer, reason, totalBalance } = data;

        const queryStr = 'INSERT INTO patients (patient_firstName, patient_lastName, patient_middleName, patient_birthDate, patient_age, patient_sex, patient_nickname, patient_nationality, patient_religion, patient_occupation, patient_referrer, patient_reason, patient_totalBal) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)';
        const values = [firstName, lastName, middleName, birthDate, age, sex, nickname, nationality, religion, occupation, referrer, reason, totalBalance];
        
        // execute
        try {
            // console.log('transaction:', transaction);
            const [result] = await pool.query(queryStr, values);
            const newPatient = result.insertId;
            if (newPatient) {
                console.log('Patient created successfully from model', newPatient);
                return newPatient;
            } else {
                throw new Error('Error creating patient');
            }
        } catch (error) {
            console.log('Error creating patient from model', error);
            throw error;
        }
    }

    static async getPatients() {
        const queryStr = 'SELECT * FROM patients';

        try {
            const [patients] = await pool.query(queryStr);
            return patients;
        } catch (error) {
            console.log('Error getting patients from model', error);
            throw error;
        }
    }

    static async getPatientByName(data) {
        const {firstName, lastName} = data;

        const queryStr = 'SELECT * FROM patients WHERE patient_firstName = ?';
        const values = [firstName, lastName];

        const [searchedPatients] = await pool.query(queryStr, values);
        return searchedPatients;
    } 

    static async getPatientById(id) {
        const queryStr = 'SELECT * FROM patients WHERE patient_id = ?';
        try {
            const [patient] = await pool.query(queryStr, [id]);
            if (patient.length === 0) {
                throw new Error('Patient data not found');
            } else {
                return patient;
            }
            
        } catch (error) {
            console.log('Error getting patient from model', error);
            throw error;
        }
    }

    static async getAllPatients() {
        const queryStr = 'SELECT * FROM patients';
        try {
            const [patients] = await pool.query(queryStr);
            return patients;
        } catch (error) {
            console.log('Error getting patients from model', error);
            throw error;
        }
    }

    // UPDATE

    static async updatePatient(data) {
        const { firstName, lastName, middleName, birthDate, age, sex, nickname, nationality, religion, occupation, reason, totalBalance, patient_id } = data;

        const queryStr = `
        UPDATE patients
        SET 
            patient_firstName = ?,
            patient_lastName = ?,
            patient_middleName = ?,
            patient_birthDate = ?,
            patient_age = ?,
            patient_sex = ?,
            patient_nickname = ?,
            patient_nationality = ?,
            patient_religion = ?,
            patient_occupation = ?,
            patient_reason = ?,
            patient_totalBal = ?
        WHERE patient_id = ?`;

    const values = [firstName, lastName, middleName, birthDate, age, sex, nickname, nationality, religion, occupation, reason, totalBalance, patient_id];
        // execute
        try {
            const [result] = await pool.query(queryStr, values);
            if (result.affectedRows === 0) {
                console.log('no updates in patient, does not exist');
                throw new Error('Patient does not exist, ');
            }

            console.log('Patient updated successfully from model');
            return {
                affectedRows: result.affectedRows,
                message: 'Patient updated successfully',
                patient_id: patient_id
            }
        } catch (error) {
            console.log('Error updating patient from model', error);
            throw error;
        }
    }

    static async deletePatient(id) {
        try {
            const queryStr = 'DELETE FROM patients WHERE patient_id = ?';
            const [result] = await pool.query(queryStr, id);

            if (result.affectedRows === 0) {
                throw new Error('Patient not deleted, might not exist');
            } else {
                return {
                    message: 'Patient deleted successfully',
                    affectedRows: result.affectedRows,
                    patient_id: id
                }
            }
        } catch (error) {
            console.log('Error deleting patient from model', error);
            throw error;
        }
    }
}

File: dentsys_client/lib/models/patient_model.dart
class Patient {
  final int? id;
  final String firstName;
  final String lastName;
  final String middleName;
  final String? nickname;
  final String birthDate;
  final int age;
  final String? sex;
  final String nationality;
  final String religion;
  final String occupation;
  final String reason;
  final String? balance;
  final String? referrer;
  final String? parentName;
  final String? parentOccupation;

  Patient({
    this.id,
    required this.firstName,
    required this.lastName,
    required this.middleName,
    this.nickname,
    required this.birthDate,
    required this.age,
    required this.sex,
    required this.nationality,
    required this.religion,
    required this. occupation,
    required this.reason,
    this.balance,
    this.referrer,
    this.parentName,
    this.parentOccupation
  });

  @override
  String toString() {
    var fullName = '$firstName ' '$lastName';
    return fullName;
  }

  Map<String, dynamic> toJson() {
    return {
      "firstName": firstName,
      "lastName": lastName,
      "middleName": middleName,
      "nickname": nickname ?? "",
      "birthDate": birthDate,
      "age": age,
      "sex": sex,
      "nationality": nationality,
      "religion": religion,
      "occupation": occupation,
      "reason": reason,
      "balance": balance ?? "",
      "referrer": referrer ?? "",
      "parentName": parentName ?? "",
      "parentOccupation": parentOccupation ?? ""
    };
    // remember: balance temporarily removed
  }

  factory Patient.fromJson(Map<String, dynamic> json){
    // print ('received in fromJson patient: $json');
    final patientData = json['newPatient'] != null ? json['newPatient'][0] : json; // if 'newPatient' exists in json response, use it, else use json 
    // print('patient in fromJson: $patientData');
    // NOTE: newPatient is for add-patient, patient is for get-patient (next time be consistent in sending response data from backend)
    return Patient(
      id: patientData['patient_id'],
      firstName: patientData['patient_firstName'],
      lastName: patientData['patient_lastName'],
      middleName: patientData['patient_middleName'],
      nickname: patientData['patient_nickname'],
      birthDate: patientData['patient_birthdate'],
      age: patientData['patient_age'] is String 
        ? int.tryParse(patientData['patient_age']) ?? 0 
        : patientData['patient_age'] as int,
      sex: patientData['patient_sex'],
      nationality: patientData['patient_nationality'],
      religion: patientData['patient_religion'],
      occupation:  patientData['patient_occupation'],
      reason: patientData['patient_reason'],
      balance: patientData['patient_totalBal'],
      parentName: patientData['patient_parentName'],
      parentOccupation: patientData['patient_parentOccupation'],
      referrer: patientData['patient_referrer']
    );
    // remember: balance temporarily removed
  }
}
